src/core/tsi/alts/crypt/aes_gcm.cc

```diff
@@ -182,38 +182,95 @@ static void aes_gcm_mask_nonce(uint8_t* dst, const uint8_t* nonce,
   memcpy(dst + sizeof(nonce1), &nonce2, sizeof(nonce2));
 }
 
-static grpc_status_code aes_gcm_derive_aead_key(uint8_t* dst,
-                                                const uint8_t* kdf_key,
-                                                const uint8_t* kdf_counter) {
-  unsigned char buf[EVP_MAX_MD_SIZE];
-  unsigned char ctr = 1;
+grpc_status_code hkdf_derive_secret(uint8_t* out_key, size_t out_size,
+                                     GsecHashFunction hash_function, const uint8_t* prk,
+                                     size_t prk_size, const uint8_t* info,
+                                     size_t info_size) {
+  const EVP_MD* digest;
+  switch (hash_function) {
+    case SHA256_hash_function:
+      digest = EVP_sha256();
+      break;
+    case SHA384_hash_function:
+      digest = EVP_sha384();
+      break;
+  }
+  const size_t digest_size = EVP_MD_size(digest);
+  uint8_t buf[EVP_MAX_MD_SIZE];
+  size_t n, done = 0;
+  unsigned i;
+  n = (out_size + digest_size - 1) / digest_size;
+  if (out_size + digest_size < out_size || n > 255) {
+    return GRPC_STATUS_INTERNAL;
+  }
 #if OPENSSL_VERSION_NUMBER < 0x10100000L
   HMAC_CTX hmac;
   HMAC_CTX_init(&hmac);
-  if (!HMAC_Init_ex(&hmac, kdf_key, kKdfKeyLen, EVP_sha256(), nullptr) ||
-      !HMAC_Update(&hmac, kdf_counter, kKdfCounterLen) ||
-      !HMAC_Update(&hmac, &ctr, 1) || !HMAC_Final(&hmac, buf, nullptr)) {
+  if (!HMAC_Init_ex(&hmac, prk, prk_size, digest, nullptr)) {
     HMAC_CTX_cleanup(&hmac);
     return GRPC_STATUS_INTERNAL;
   }
-  HMAC_CTX_cleanup(&hmac);
+  for (i = 0; i < n; i++) {
```

Comment:
Can you create a function that takes an HMAC_CTX* and executes the loop (minus HMAC_CTX_cleanup/HMAC_CTX_free)? Then the two versions can call this function and clean up on their own.

Reply:
Done.